public/app/features/dashboard/components/PanelEditor/PanelEditorTabs.tsx

```diff
@@ -16,55 +16,79 @@ interface PanelEditorTabsProps {
   onChangeTab: (tab: PanelEditorTab) => void;
 }
 
-export const PanelEditorTabs: React.FC<PanelEditorTabsProps> = ({ panel, dashboard, tabs, onChangeTab }) => {
-  const styles = getPanelEditorTabsStyles();
-  const activeTab = tabs.find(item => item.active)!;
+export class PanelEditorTabs extends PureComponent<PanelEditorTabsProps> {
+  componentDidMount() {
+    this.props.panel.on(CoreEvents.queryAdded, this.triggerForceUpdate);
+    this.props.panel.on(CoreEvents.queryRemoved, this.triggerForceUpdate);
+    this.props.panel.on(CoreEvents.alertAdded, this.triggerForceUpdate);
+    this.props.panel.on(CoreEvents.alertRemoved, this.triggerForceUpdate);
+    this.props.panel.on(CoreEvents.transformationAdded, this.triggerForceUpdate);
+    this.props.panel.on(CoreEvents.transformationRemoved, this.triggerForceUpdate);
+  }
 
-  const getCounter = useCallback(
-    (tab: PanelEditorTab) => {
-      switch (tab.id) {
-        case PanelEditorTabId.Query:
-          return panel.targets.length;
-        case PanelEditorTabId.Alert:
-          return panel.alert ? 1 : 0;
-        case PanelEditorTabId.Transform:
-          const transformations = panel.getTransformations() ?? [];
-          return transformations.length;
-      }
+  componentWillUnmount() {
+    this.props.panel.off(CoreEvents.queryAdded, this.triggerForceUpdate);
+    this.props.panel.off(CoreEvents.queryRemoved, this.triggerForceUpdate);
+    this.props.panel.off(CoreEvents.alertAdded, this.triggerForceUpdate);
+    this.props.panel.off(CoreEvents.alertRemoved, this.triggerForceUpdate);
+    this.props.panel.off(CoreEvents.transformationAdded, this.triggerForceUpdate);
+    this.props.panel.off(CoreEvents.transformationRemoved, this.triggerForceUpdate);
```

Comment:
@peterholmberg noticed that e2e tests were complaining with this:
```
(node:20782) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 preprocessor:close listeners added to [EventEmitter]. Use emitter.setMaxListeners() to increase limit
(node:20782) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 preprocessor:close listeners added to [EventEmitter]. Use emitter.setMaxListeners() to increase limit
```

I've checked and this is already on master @peterholmberg so this PR hasn't introduce it.

But when I debug this PR and reach `componentWillUnmount` the events array in the emitter is already empty. Could be because Panel has already been destroyed or the props.panel is a different reference? https://github.com/grafana/grafana/blob/17a1e78754a036b2671437e9b59b0734385c74e4/public/app/features/dashboard/state/PanelModel.ts#L454

Not sure what the best way is to register/unregister emitter this without memory leaks in this case, maybe @torkelo @dprokop has suggestions?

Reply:
If this is the case maybe we should store the `unsubscribe` functions for each subscription within this panel and run those on unmount instead of the `off`.

Reply:
Didn't know it returned an `unsubscribe` function, that would be a better approach indeed!